## Аналіз A/B тесту

Маємо проаналізувати дані A/B тесту в популярній [грі Cookie Cats](https://www.facebook.com/cookiecatsgame). Це класична гра-головоломка в стилі «з’єднай три», де гравець повинен з’єднати плитки одного кольору, щоб очистити дошку та виграти рівень. На дошці також зображені співаючі котики :)

Під час проходження гри гравці стикаються з воротами, які змушують їх чекати деякий час, перш ніж вони зможуть прогресувати або зробити покупку в додатку. У цьому блоці завдань ми проаналізуємо результати A/B тесту, коли перші ворота в Cookie Cats було переміщено з рівня 30 на рівень 40. Зокрема, ми проаналізуємо вплив на утримання (retention) гравців. Тобто хочемо зрозуміти чи переміщення воріт на 10 рівнів пізніше якимось чином вплинуло на те, що користувачі перестають грати в гру раніше чи пізніше з точки зору кількості їх днів з моменту встановлення гри.

Будемо працювати з даними з файлу `cookie_cats.csv`. Змінні в даних наступні:

- userid - унікальний номер, який ідентифікує кожного гравця.
- version - чи потрапив гравець в контрольну групу (gate_30 - ворота на 30 рівні) чи тестову групу (gate_40 - ворота на 40 рівні).
- sum_gamerounds - кількість ігрових раундів, зіграних гравцем протягом першого тижня після встановлення
- retention_1 - чи через 1 день після встановлення гравець повернувся і почав грати?
- retention_7 - чи через 7 днів після встановлення гравець повернувся і почав грати?

Коли гравець встановлював гру, його випадковим чином призначали до групи gate_30 або gate_40.

1. Зчитайте дані АВ тесту у змінну `df` та виведіть середнє значення показника показник `retention_7` (утримання на 7 день) по версіям гри. Сформулюйте гіпотезу: яка версія дає краще утримання через 7 днів після встановлення гри?

In [17]:
import pandas as pd
import numpy as np
#df = pd.read_csv('data/cookie_cats.csv')
df = pd.read_csv('/Users/julia/data/data_statistics/cookie_cats.csv')
df

userid  version  sum_gamerounds  retention_1  retention_7
0          116  gate_30               3        False        False
1          337  gate_30              38         True        False
2          377  gate_40             165         True        False
3          483  gate_40               1        False        False
4          488  gate_40             179         True         True
...        ...      ...             ...          ...          ...
90184  9999441  gate_40              97         True        False
90185  9999479  gate_40              30        False        False
90186  9999710  gate_30              28         True        False
90187  9999768  gate_40              51         True        False
90188  9999861  gate_40              16        False        False

[90189 rows x 5 columns]

In [19]:
grouped = df.groupby("version")['retention_7'].mean()
grouped

version
gate_30    0.190201
gate_40    0.182000
Name: retention_7, dtype: float64

# Висновки
Ці середні значення представляють частку гравців, які повернулися до гри через 7 днів після її встановлення, для кожної з версій гри:

gate_30 (0.190201): приблизно 19% гравців, які мали ворота на рівні 30, повернулися до гри через 7 днів після встановлення.

gate_40 (0.182000): приблизно 18.2% гравців, які мали ворота на рівні 40, повернулися до гри через 7 днів після встановлення.
На основі цих середніх можна зробити попередній висновок, що утримання на 7 день трохи вище в контрольній групі (ворота на рівні 30).

2. Перевірте з допомогою z-тесту аналогічно до прикладу в лекції, чи дає якась з версій гри кращий показник `retention_7` на рівні значущості 0.05. Обчисліть також довірчі інтервали для двох вибірок. Виведіть результат у форматі:
```
z statistic: ...
p-value: ...
Довірчий інтервал 95% для групи control: [..., ...]
Довірчий інтервал 95% для групи treatment: [..., ...]
```
де замість `...` - обчислені значення. В якості висновка дайте відповідь на два питання:  
    1. чи є статистична значущою різниця між поведінкою користувачів у різних версіях гри?   
    2. чи перетинаються довірчі інтервали утримання користувачів з різних версій гри? Про що це каже?  
    
Зверніть увагу, в такому і схожому завданнях ми використовуєм `proportion` Z-тест. Це тому що в нас залежна змінна має бінарне значення (повернеться аби ні користувач, чи клікне або ні користувач в інших ситуаціях - всього два можливих значення в змінної: 0/1, True/False ). Якщо б ми вимірювали скажімо чи є стат. значущою різниця між вагою чоловіків і жінок в певній вибірці, ми б використовувавли функцію `statsmodels.stats.ztest`, бо залежна змінна `вага` є неперервною (тип float, замість типу int чи bool і тільки двох можливих значень).

In [26]:
import numpy as np
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

# Кількість гравців у кожній групі
n_gate_30 = df[df['version'] == 'gate_30']['retention_7'].count()
n_gate_40 = df[df['version'] == 'gate_40']['retention_7'].count()

# Кількість гравців, які повернулися на 7 день
success_gate_30 = df[df['version'] == 'gate_30']['retention_7'].sum()
success_gate_40 = df[df['version'] == 'gate_40']['retention_7'].sum()

# Пропорції
p_gate_30 = success_gate_30 / n_gate_30
p_gate_40 = success_gate_40 / n_gate_40

# Z-тест для пропорцій
count = np.array([success_gate_30, success_gate_40])
nobs = np.array([n_gate_30, n_gate_40])
z_stat, p_value = proportions_ztest(count, nobs)

# Довірчі інтервали для кожної групи
ci_gate_30 = proportion_confint(success_gate_30, n_gate_30, alpha=0.05)
ci_gate_40 = proportion_confint(success_gate_40, n_gate_40, alpha=0.05)


# Виводимо результати
print(f"z statistic: {z_stat:.4f}")
print(f"p-value: {p_value:.4f}")
print(f"Довірчий інтервал 95% для групи gate_30: [{ci_gate_30[0]:.4f}, {ci_gate_30[1]:.4f}]")
print(f"Довірчий інтервал 95% для групи gate_40: [{ci_gate_40[0]:.4f}, {ci_gate_40[1]:.4f}]")


z statistic: 3.1644
p-value: 0.0016
Довірчий інтервал 95% для групи gate_30: [0.1866, 0.1938]
Довірчий інтервал 95% для групи gate_40: [0.1785, 0.1855]


# Висновки
p-value (0.0016) менше рівня значущості (0.05). Це свідчить про те, що різниця в утриманні користувачів між версіями gate_30 і gate_40 є статистично значущою.

довірчі інтервали для gate_30 [0.1866, 0.1938] і gate_40 [0.1785, 0.1855] не перетинаються. Це вказує на те, що утримання користувачів у цих двох групах дійсно відрізняється, і це підтверджує результати z-тесту.

Це означає, що переміщення воріт вплинуло на утримання гравців, і версія з ворота на рівні 30 має кращі результати.

3. Є ще один тип тестів, який використовується для бінарної метрики як от "зробить юзер дію, чи ні" - тест [**Хі-квадрат**](https://www.bmj.com/about-bmj/resources-readers/publications/statistics-square-one/8-chi-squared-tests) (ще одне [пояснення](https://www.scribbr.com/statistics/chi-square-tests/) тесту з прикладами). В нього інші гіпотези Н0 і Н1 на відміну від z- та t-тестів. А також цей тест можна використовувати, якщо в нас більше за 2 досліджувані групи, тобто в нас не А/В тест, а А/B/C/D, наприклад.  

В **z- та t-тестах** (які відрізняються тим, що ми в першому не знаємо дисперсію генеральної сукупності, але якщо в нас великий набір даних, то ці два тести дають дуже схожі результати) **ми перевіряємо, чи є різниця у середніх показниках по групам користувачів**.  

А в **тесті Хі-квадрат ми перевіряємо чи є звʼязок між групою користувача і тим, чи він зробить цікаву нам дію**. Це ніби дослідження одного і того самого, але дещо різними способами. Для перевірки, можна виконувати кілька тестів (особливо, якщо один дає якийсь непереконливий результат типу р-значення 0.07 - наче і fail to regect H0 на рівні стат значущості 5%, але цікаво, що скажуть інші тести), тож, зробимо і ми тест хі-квадрат та порівняємо його результат з z-тестом.

Про різницю між тестами можна почитати ще [тут](https://stats.stackexchange.com/a/178860) - це просто пояснення користувача стековерфлоу, але там розумні люди сидять.

Для проведення хі-квадрат тесту скористаємось функцією з `scipy.stats` `chi2_contingency` для обчислення статистики хі-квадрат і р-значення для перевірки конкретної гіпотези. У цю функцію вам треба передати таблицю 2х2: кількість випадків для кожної версії гри і значення `retention_7`.

**Задача**: виконайте тест хі-квадрат на рівні значущості 5% аби визначити, чи є залежність між версією гри та тим, чи зайде гравець на 7ий день після встановлення гри.
Тут гіпотези наступні
- Н0: значення retention_7 не залежить від версії гри
- Н1: є залежність між версією гри і значенням retention_7

Виведіть p-значення та зробіть висновок.


In [31]:
import pandas as pd
from scipy.stats import chi2_contingency

# Створюємо таблицю 2x2 для хі-квадрат тесту
# Розраховуємо кількість гравців, які повернулися (True) та не повернулися (False) для кожної версії
table = pd.crosstab(df['version'], df['retention_7'])

# Виконуємо тест хі-квадрат
chi2_stat, p_value, dof, expected = chi2_contingency(table)

# Виводимо результати
print(f"p-value: {p_value:.4f}")


alpha = 0.05
if p_value < alpha:
    print("Відхиляємо нульову гіпотезу (H0): існує залежність між версією гри і значенням retention_7.")
else:
    print("Не відхиляємо нульову гіпотезу (H0): значення retention_7 не залежить від версії гри.")


p-value: 0.0016
Відхиляємо нульову гіпотезу (H0): існує залежність між версією гри і значенням retention_7.
